### PR TITLE
Add function to load Qiskit operators

### PR DESCRIPTION
**Pull request by Mandrenkov**
**Context:**

The next release of the PennyLane-Qiskit plugin will include PennyLaneAI/pennylane-qiskit#401 which features a new function for converting Qiskit [SparsePauliOp](https://docs.quantum.ibm.com/api/qiskit/qiskit.quantum_info.SparsePauliOp) instances into PennyLane operators.

**Description of the Change:**

* Added a `from_qiskit_op()` function to the IO module for loading Qiskit operators.

In a nutshell,

```python
>>> import pennylane as qml
>>> from qiskit.quantum_info import SparsePauliOp
>>> qiskit_op = SparsePauliOp(["II", "XY"])
>>> qml.from_qiskit_op(qiskit_op)
I(0) + X(1) @ Y(0)
```

**Benefits:**

* Qiskit operators can be directly converted into PennyLane operators.

**Possible Drawbacks:**

None.

**Related GitHub Issues:**

None.

### CODE REVIEW

File: pennylane/io.py [hunk 1]
```diff
@@ -146,13 +156,99 @@ def circuit_loaded_qiskit_circuit():
...
+
+    >>> qiskit_op
+    SparsePauliOp(['II', 'XY'],
+                  coeffs=[1.+0.j, 1.+0.j])
+
+    To convert the ``SparsePauliOp`` into a PennyLane :class:`Operator`, use:
+
+    >>> import pennylane as qml
+    >>> qml.from_qiskit_op(qiskit_op)
+    I(0) + X(1) @ Y(0)
```
Comment (trbromley) [inline pennylane/io.py:205]: Minor point, but I kind of expected the order to be swapped:
```
I(0) + Y(0) @ X(1)
```

File: pennylane/io.py [hunk 2]
```diff
@@ -146,13 +156,99 @@ def circuit_loaded_qiskit_circuit():
...
+
+    Returns:
+        Operator: The equivalent PennyLane operator.
+
+    .. note::
+
+        The wire ordering convention differs between PennyLane and Qiskit: PennyLane wires are
+        enumerated from left to right, while the Qiskit convention is to enumerate from right to
+        left. A ``SparsePauliOp`` term defined by the string ``"XYZ"`` applies ``Z`` on wire 0,
+        ``Y`` on wire 1, and ``X`` on wire 2.
```
Comment (trbromley) [inline pennylane/io.py:183]: How do we know this is the case? Is it reported somewhere in the Qiskit docs?

**Review by trbromley — COMMENTED** [review]
Thanks @Mandrenkov!

File: pennylane/io.py (hunk 2 above)
Comment (lillian542) [reply pennylane/io.py:183]: Passing the string "XYZ" creates a `qiskit.quantum_info.Pauli` object in Qiskit, and its described in the docs for that here: https://docs.quantum.ibm.com/api/qiskit/qiskit.quantum_info.Pauli

Comment (lillian542) [reply pennylane/io.py:183]: Documentation for `SparsePauliOp` mentions that the Paulis are a `PauliList`, and documentation for the `PauliList` mentions that the elements are `Paulis`. I don't know of a place where it's explained "closer" to the SparsePauliOps class. Should we make this text more specific to describing the wire ordering of a Pauli and provide the link?

File: pennylane/io.py [hunk 3]
```diff
@@ -146,13 +156,99 @@ def circuit_loaded_qiskit_circuit():
...
-                "You may need to restart your kernel or environment after installation. "
-                "If you have any difficulties, you can reach out on the PennyLane forum at "
-                "https://discuss.pennylane.ai/c/pennylane-plugins/pennylane-qiskit/"
-            ) from e
+            raise RuntimeError(_MISSING_QISKIT_PLUGIN_MESSAGE) from e
+        raise e
+
+
+def from_qiskit_op(qiskit_op, params=None, wires=None):
+    """Loads Qiskit SparsePauliOp objects by using the converter in the PennyLane-Qiskit plugin.
```
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:164]: > Can we provide a link to `SparsePauliOp` in the Qiskit docs.

Sure, can do! I didn't include it before to be consistent with the `from_qiskit()` docstring.

[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:164]: > Also, I think we don't need to mention that this comes from the PL-Qiskit plugin.

Actually, this is the documentation convention for all of the `from_XXX()` convenience functions in this file so I would prefer to keep things consistent unless we have a good reason not to.

File: pennylane/io.py [hunk 4]
```diff
@@ -146,13 +156,99 @@ def circuit_loaded_qiskit_circuit():
...
+
+    Args:
+        qiskit_op (qiskit.quantum_info.SparsePauliOp): the SparsePauliOp to be converted
```
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:167]: Sure, I can adjust those. Should I do the same for `QuantumCircuit` in the function above?

File: pennylane/io.py [hunk 5]
```diff
@@ -146,13 +156,99 @@ def circuit_loaded_qiskit_circuit():
...
+        params (Any): optional assignment of coefficient values for the SparsePauliOp; see the
```
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:168]: Not quite: the type is either `None`, `Sequence`, or `Mapping`, [or more specifically](https://docs.quantum.ibm.com/api/qiskit/qiskit.quantum_info.SparsePauliOp#assign_parameters),
```python
None|Mapping[Parameter, complex|ParameterExpression] | Sequence[complex|ParameterExpression]
```
I opted to leave this as `Any` because we don't really care about the type (we just forward this argument to `assign_parameters()`), to keep things simple, and to insulate ourselves from future changes made by Qiskit.

File: pennylane/io.py (hunk 1 above)
Comment (Mandrenkov) [reply pennylane/io.py:205]: I agree although I think changing that behaviour would fall outside the scope of this PR.

Is there an existing Story to track this, @trbromley?

File: pennylane/io.py (hunk 2 above)
Comment (trbromley) [reply pennylane/io.py:183]: You're right! I'm sad though because it just makes things more confusing :disappointed: 
This is how I checked:
```python
from qiskit.quantum_info import SparsePauliOp
from qiskit import QuantumCircuit
from qiskit_aer import AerSimulator
from qiskit.opflow import CircuitSampler
from qiskit.opflow import StateFn, PauliExpectation, CircuitSampler
from qiskit_aer import AerSimulator
from qiskit.opflow import CircuitStateFn, PauliSumOp
 
op = SparsePauliOp(["XZ"], [1])

state = QuantumCircuit(2)
state.h(0)

simulator = AerSimulator(method="statevector", shots=100)

# opflow stuff
opflow_op = PauliSumOp(op)
opflow_state = CircuitStateFn(state)
measurable_expression = StateFn(opflow_op, is_measurement=True).compose(opflow_state)
expectation = PauliExpectation().convert(measurable_expression)
simulator = AerSimulator(method="statevector", shots=100)
circuit_sampler = CircuitSampler(simulator).convert(expectation)
 
expectation_value = circuit_sampler.eval().real
```
The result is: `0.0`. Compared to PL:
```python
import pennylane as qml

dev = qml.device("default.qubit")

@qml.qnode(dev)
def f():
    qml.Hadamard(0)
    return qml.expval(qml.PauliX(0) @ qml.PauliZ(1)), qml.expval(qml.PauliX(1) @ qml.PauliZ(0))

f()
```
We get `(tensor(1., requires_grad=True), tensor(0., requires_grad=True))`, which agrees with `X` being on qubit `1` not `0`.

Comment (Mandrenkov) [reply pennylane/io.py:183]: This is mentioned in the ["String representation"](https://docs.quantum.ibm.com/api/qiskit/qiskit.quantum_info.Pauli) section of the documentation for `Pauli`:

> In the string representation qubit-0 corresponds to the right-most Pauli character, and qubit-($n$−1) to the left-most Pauli character. For example `'XYZ'` represents  $X$ ⊕ $X$ ⊕ $Z$ with `'Z'` on qubit-0, `'Y'` on qubit-1, and `'X'` on qubit-2.

Should I include a link to it?

File: pennylane/io.py [hunk 6]
```diff
@@ -146,13 +156,99 @@ def circuit_loaded_qiskit_circuit():
...
+        SparsePauliOp(['XYZ'],
+              coeffs=[1.+0.j])
+        >>> qml.from_qiskit_op(wired_qiskit_op, wires=[3, 5, 7])
+        Y(5) @ Z(3) @ X(7)
+    """
+    try:
+        return load(qiskit_op, format="qiskit_op", params=params, wires=wires)
+    except ValueError as e:
+        if e.args[0].split(".")[0] == "Converter does not exist":
+            raise RuntimeError(_MISSING_QISKIT_PLUGIN_MESSAGE) from e
```
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:251]: Perhaps, but that would be inconsistent with `from_qiskit()` and we would still need to deal with dynamically importing the package and checking if that succeeds.

**Comment by Mandrenkov** [comment]
Thanks for the review, @trbromley!

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -10,6 +10,9 @@
   function fails because the Qiskit converter is missing.
   [(#5218)](https://github.com/PennyLaneAI/pennylane/pull/5218)
 
+* Qiskit operators can be converted into PennyLane operators using `qml.from_qiskit_op`.
```
Comment (lillian542) [inline doc/releases/changelog-dev.md:13]: For clarity, could we say "A Qiskit `SparsePauliOp` can be converted" instead of "Qiskit operators"?

File: pennylane/io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/io.py:164]: > Sure, can do! I didn't include it before to be consistent with the from_qiskit() docstring.

Ah, we can break convention with `from_qiskit()` as we'll be revisiting that with the docs improvements next week.

[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/io.py:164]: > Actually, this is the documentation convention for all of the from_XXX() convenience functions in this file so I would prefer to keep things consistent unless we have a good reason not to.

Also don't mind breaking convention here. My motivation is that if you're a user who sees `from_qiskit_op`, you probably don't care too much that the functionality lives in the PL-Qiskit plugin. Seems more like an implementation detail.

File: pennylane/io.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/io.py:167]: Thanks! Yes would be nice to update for `QuantumCircuit` in `from_qiskit`, but also open to that being in a separate PR next week when we have more time.

File: pennylane/io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/io.py:164]: Also, will we be able to backport these docs updates to the docstrings in the PL Qiskit plugin?

File: pennylane/io.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/io.py:168]: Thanks, got it! Ok `Any` is good.

File: pennylane/io.py (hunk 2 above)
Comment (trbromley) [reply pennylane/io.py:183]: Right, yes would be nice to link to there in our warning.

File: pennylane/io.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/io.py:251]: I just have a minor preference to avoid entrypoints unless we need to use entrypoints. Entrypoints can cause confusion for devs (e.g., having to reinstall their editable PL version) and also for users on cloud notebooks (might have to restart the kernel multiple times).

File: pennylane/io.py (hunk 2 above)
Comment (lillian542) [reply pennylane/io.py:183]: It definitely does, and we have to account for this various places in the plugin - for example, we have to reverse the order of the samples they return. We may want to think about how to document this in general for people who might be coming to PennyLane from Qiskit.

Wrt the conversation a few days ago, this is also a good reason to check the devices vs default qubit in the test suite, because this has slipped under the radar before; most recently we fixed a related issue that had been present for quite some time [in the Sampler device](https://github.com/PennyLaneAI/pennylane-qiskit/pull/357#discussion_r1421093292).

File: pennylane/io.py (hunk 1 above)
Comment (trbromley) [reply pennylane/io.py:205]: Yes ok not worth blocking this PR. I'm not sure if we have a story for this, I guess it comes back to us not being able to always impose an order on the wires used.

Comment (lillian542) [reply pennylane/io.py:205]: to add - this is an idiosyncrasy of calling `simplify`, i.e. 

```
>>> ops = [qml.Y(0), qml.X(1)]
>>> qml.prod(*ops)
Y(0) @ X(1)
```

```
>>> qml.prod(*ops).simplify()
X(1) @ Y(0)
```

File: pennylane/io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:164]: > probably don't care too much that the functionality lives in the PL-Qiskit plugin. Seems more like an implementation detail.

@trbromley On one hand, I agree, but on the other hand, this informs the user that the PL-Qiskit plugin is required.

File: pennylane/io.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:251]: That is fair although I'm not sure using dynamic imports solves that problem (unless we force-reload the import every time). Do you think it's worth hosting a discussion with the broader team about these approaches?

**Review by lillian542 — COMMENTED** [review]
Looking good to me, and all seems to be working well in a Jupyter notebook. Just let me know when its ready for a final review!

File: doc/releases/changelog-dev.md (hunk 7 above)
Comment (Mandrenkov) [reply doc/releases/changelog-dev.md:13]: Sure! I opted for "Qiskit operators" to better match the function name but I'm OK with being more specific too.

File: pennylane/io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/io.py]: It's a good point. Let's merge this in for now. Next week we could think about instead having it as a separate note rather than in the first line of the docstring. The note can give a few more details, like installation instructions.

File: pennylane/io.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply pennylane/io.py:254]: After discussing this offline, we agreed to leave the usage of entry points for now and revisit the matter later.

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -9,6 +9,8 @@
 * An error message provides clearer instructions for installing PennyLane-Qiskit if the `qml.from_qiskit`
   function fails because the Qiskit converter is missing.
   [(#5218)](https://github.com/PennyLaneAI/pennylane/pull/5218)
+* A Qiskit `SparsePauliOp` can be converted into a PennyLane `Operator` using `qml.from_qiskit_op`.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:12]:
```suggestion

* A Qiskit `SparsePauliOp` can be converted into a PennyLane `Operator` using `qml.from_qiskit_op`.
```

File: pennylane/io.py [hunk 9]
```diff
@@ -146,13 +156,102 @@ def circuit_loaded_qiskit_circuit():
...
-                "You may need to restart your kernel or environment after installation. "
-                "If you have any difficulties, you can reach out on the PennyLane forum at "
-                "https://discuss.pennylane.ai/c/pennylane-plugins/pennylane-qiskit/"
-            ) from e
+            raise RuntimeError(_MISSING_QISKIT_PLUGIN_MESSAGE) from e
+        raise e
+
+
+def from_qiskit_op(qiskit_op, params=None, wires=None):
+    """Loads Qiskit `SparsePauliOp <https://docs.quantum.ibm.com/api/qiskit/qiskit.quantum_info.SparsePauliOp>`_
```
Comment (trbromley) [inline pennylane/io.py:164]: Do we need single underscores or double?

**Review by trbromley — APPROVED** [review]
Thanks @Mandrenkov!!

File: pennylane/io.py (hunk 9 above)
Comment (Mandrenkov) [reply pennylane/io.py:164]: It looks like [single underscores should be fine](https://docutils.sourceforge.io/docs/user/rst/quickref.html#hyperlink-targets). I'm not sure where the `__` convention originated.

**Review by lillian542 — APPROVED** [review]
Looks good to me, thanks @Mandrenkov! 🚀
